src/runtime/cuda/cuda_hardware_manager.cpp

```diff
@@ -56,6 +56,27 @@ cuda_hardware_manager::cuda_hardware_manager(hardware_platform hw_platform)
     _devices.emplace_back(dev);
   }
 
+  for (int dev = 0; dev < num_devices; ++dev) {
+    cudaSetDevice(dev);
```

Comment:
Maybe it might be cleaner to use `hip/cuda_device_manager::get().activate_device(dev);` here?

This is the pattern we usually use to set the active device. It caches the active device and only sets it if needed. Setting it directly might be a little error prone, although I suspect it's currently fine since the `hip/cuda_device_manager` is likely only first used (and thus initialized) after this point.